### PR TITLE
fix(rest): remove trailing slash for routing match

### PR DESCRIPTION
**Pull request by raymondfeng**
Handle trailing slash for urls.

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

**Comment by raymondfeng** [comment]
Please note the explorer redirect test is failing as we now treat `/explorer/` the same as `/explorer`. What's the best practice for trailing `/`? Google seems to have this view - https://webmasters.googleblog.com/2010/04/to-slash-or-not-to-slash.html.

What do you think? Should we fix the explorer redirect for `/explorer/` as it's not needed now?

**Comment by raymondfeng** [comment]
I added commit https://github.com/strongloop/loopback-next/pull/2355/commits/402a80db2fea254653c10a8f40b41ac7e9ccf6d7 to make sure all tests are passing.

File: packages/rest-explorer/src/__tests__/acceptance/rest-explorer.acceptance.ts [hunk 1]
```diff
@@ -36,11 +36,12 @@ describe('API Explorer (acceptance)', () => {
         .expect(/<title>LoopBack API Explorer/);
     });
 
-    it('redirects from "/explorer" to "/explorer/"', async () => {
+    it('exposes API Explorer at "/explorer"', async () => {
```
Comment (bajtos) [inline packages/rest-explorer/src/__tests__/acceptance/rest-explorer.acceptance.ts]: This redirect is important because the way how relative paths are resolved in the browser. When the SPA UI is served from `/explorer`, a path like `src="swagger-ui.js"` is resolved to `/swagger-ui.js` instead of `/explorer/swagger-ui.js`.

I don't see how your patch is fixing this problem.

Can you revert the changes in REST Explorer please?

File: packages/rest/src/router/router-base.ts [hunk 2]
```diff
@@ -18,12 +18,29 @@ export abstract class BaseRouter implements RestRouter {
...
 
-  protected getKeyForRoute(route: RouteEntry) {
-    const path = route.path.startsWith('/') ? route.path : `/${route.path}`;
-    const verb = route.verb.toLowerCase() || 'get';
+  /**
+   * Build a key for verb+path
+   * @param verb HTTP verb/method
+   * @param path URL path
+   */
+  private getKey(verb: string, path: string) {
```
Comment (bajtos) [inline packages/rest/src/router/router-base.ts]: Since this method is not using anything from the class instance (no `this`), can you move it outside of the class into a regular function?

**Comment by raymondfeng** [comment]
@bajtos Without https://github.com/strongloop/loopback-next/commit/402a80db2fea254653c10a8f40b41ac7e9ccf6d7, CI is failing. See https://travis-ci.org/strongloop/loopback-next/jobs/490404103. 

After we normalizing the key for route matching, `/explorer` matches `/explorer/` and vice versa. As a result, `/explorer` returns `200`.

**Comment by raymondfeng** [comment]
@bajtos I improved `rest-explorer` to honor `explorerPath` in index.html. PTAL.

File: packages/rest-explorer/src/__tests__/acceptance/rest-explorer.acceptance.ts (hunk 1 above)
Comment (raymondfeng) [reply packages/rest-explorer/src/__tests__/acceptance/rest-explorer.acceptance.ts]: Instead of depending on redirect, I updated `index.html.ejs` to honor `explorerPath` and it works now for `http://localhost:3000/explorer`.

**Comment by raymondfeng** [comment]
@bajtos PTAL

File: packages/rest-explorer/templates/index.html.ejs [hunk 3]
```diff
@@ -5,9 +5,9 @@
 <head>
   <meta charset="UTF-8">
   <title>LoopBack API Explorer</title>
-  <link rel="stylesheet" type="text/css" href="./swagger-ui.css">
-  <link rel="icon" type="image/png" href="./favicon-32x32.png" sizes="32x32" />
-  <link rel="icon" type="image/png" href="./favicon-16x16.png" sizes="16x16" />
+  <link rel="stylesheet" type="text/css" href="<%- explorerPath %>/swagger-ui.css">
```
Comment (bajtos) [inline packages/rest-explorer/templates/index.html.ejs]: Are you sure this is going to work when the app is running behind a reverse proxy like nginx?

**Comment by bajtos** [comment]
> @bajtos Without https://github.com/strongloop/loopback-next/commit/402a80db2fea254653c10a8f40b41ac7e9ccf6d7, CI is failing. See https://travis-ci.org/strongloop/loopback-next/jobs/490404103. 
>
> After we normalizing the key for route matching, `/explorer` matches `/explorer/` and vice versa. As a result, `/explorer` returns `200`.

That looks like a breaking change to me!

I agree that `/explorer` should match `/explorer/` and vice versa but **only when there is no handler for the other path registered**.

I am concerned that the behavior you are proposing here will be confusing to our users: they register handlers for two different paths, the framework happily accepts then and then only one handler is invoked for both paths.

**Comment by raymondfeng** [comment]
FYI: Express has the following [setting](http://expressjs.com/en/4x/api.html#app.set:
) to control the trialing `/` :

strict routing | Boolean | Enable strict routing. When enabled, the router treats "/foo" and "/foo/" as different. Otherwise, the router treats "/foo" and "/foo/" as the same.NOTE: Sub-apps will inherit the value of this setting. | N/A (undefined)
-- | -- | -- | --

**Comment by hacksparrow** [comment]
I think we should have an option to configure this behavior.

Btw, the the setting is simply `strict` in case of a `Router` - http://expressjs.com/en/4x/api.html#express.router.

**Comment by raymondfeng** [comment]
@bajtos @hacksparrow I extracted the fix for #2188 into a new PR - #2375 and leave the trailing slash handling in this PR, which may require more discussions.

File: packages/rest-explorer/templates/index.html.ejs (hunk 3 above)
Comment (raymondfeng) [reply packages/rest-explorer/templates/index.html.ejs]: It's the same deal with or without this change. See https://github.com/strongloop/loopback-next/pull/2293.

**Comment by raymondfeng** [comment]
Superseded by #2419
